### ISSUE TITLE
Does the werewolf work as intended?

### ISSUE THREAD

**Issue by eriktorbjorn** [Quuxplusone/Castlequest#8]
I played around with the game for a while, and the werewolf finally showed up. I tried shooting it, but missed and then he just kept following me around, taking the occasional swipe at me, without ever actually hitting me.

I don't speak Fortran, but from wwolf.f I get the impression that XLIM is the chance that he will land a fatal hit, and that this is supposed to go up every time he misses. It seems that he misses when VAL is greater than XLIM, but it seems XLIM is always 0 at this point? Is it supposed to retain its old value the next time the subroutine is called?

**Comment by Quuxplusone** [comment]
I'll take a closer look at some point for sure, but FYI, I've definitely managed to get killed by the werewolf somehow. Also, semi-spoiler, note that if you miss him once you can still try to hit him again. ;)

**Comment by eriktorbjorn** [comment]
I hadn't found any other weapon than the gun (and the hatchet, which I don't think is used as a weapon) at the time, and once I had fired the one bullet it said I didn't have any more. But maybe I missed something obvious. I was mostly focused on mapping the small part of the game I have found so far.

I tried to find information about whether the value of variables in subroutines are preserved between calls, but not being familiar with Fortran all I could figure out was that they _can_ be. But apparently they aren't for me by default. I cheated a bit to make sure the werewolf would appear early here, but I don't think that should affect the behavior I'm looking at:

```
W
0  RETURN FROM MOVE: ROOM=       5 LROOM =       3
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  19
0  ABOUT TO CALL "DES": II IS 205
0  REQUEST FOR DESCRIPTION # 205
0  You are in the dining room.
0  There is a fearsome werewolf in the room with you!
0  The werewolf attacks you with its sharp claw!
0  It just misses your neck!
0  WOLF : TXLIM:0.150

E
0  RETURN FROM MOVE: ROOM=       3 LROOM =       5
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  20
0  ABOUT TO CALL "DES": II IS 203
0  REQUEST FOR DESCRIPTION # 203
0  You're in the parlor.
0  REQUEST FOR DESCRIPTION # 610
0  There is some "HORROR HOTEL" writing paper here. 
0  REQUEST FOR DESCRIPTION # 620
0  There is an old gun here. 
0  There is a fearsome werewolf in the room with you!
0  WOLF : TXLIM:0.000

W
0  RETURN FROM MOVE: ROOM=       5 LROOM =       3
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  21
0  ABOUT TO CALL "DES": II IS 205
0  REQUEST FOR DESCRIPTION # 205
0  You are in the dining room.
0  There is a fearsome werewolf in the room with you!
0  The werewolf attacks you with its sharp claw!
0  It just misses your neck!
0  WOLF : TXLIM:0.150
```

If I understand correctly, the only reason XLIM was 0.150 here is that it was increased _after_ the check to see if the werewolf hit.

If I add `SAVE XLIM` to the subroutine, it behaves more like how I would have expected:

```
W
0  RETURN FROM MOVE: ROOM=       5 LROOM =       3
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  17
0  ABOUT TO CALL "DES": II IS 205
0  REQUEST FOR DESCRIPTION # 205
0  You are in the dining room.
0  There is a fearsome werewolf in the room with you!
0  The werewolf attacks you with its sharp claw!
0  It just misses your neck!
0  WOLF : TXLIM:0.150

0  RETURN FROM MOVE: ROOM=       3 LROOM =       5
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  18
0  ABOUT TO CALL "DES": II IS 203
0  REQUEST FOR DESCRIPTION # 203
0  You're in the parlor.
0  REQUEST FOR DESCRIPTION # 610
0  There is some "HORROR HOTEL" writing paper here. 
0  REQUEST FOR DESCRIPTION # 620
0  There is an old gun here. 
0  There is a fearsome werewolf in the room with you!
0  The werewolf attacks you with its sharp claw!
0  It just misses your neck!
0  WOLF : TXLIM:0.300

W
0  RETURN FROM MOVE: ROOM=       5 LROOM =       3
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  19
0  ABOUT TO CALL "DES": II IS 205
0  REQUEST FOR DESCRIPTION # 205
0  You are in the dining room.
0  There is a fearsome werewolf in the room with you!
0  WOLF : TXLIM:0.300

E
0  RETURN FROM MOVE: ROOM=       3 LROOM =       5
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  20
0  ABOUT TO CALL "DES": II IS 203
0  REQUEST FOR DESCRIPTION # 203
0  You're in the parlor.
0  REQUEST FOR DESCRIPTION # 610
0  There is some "HORROR HOTEL" writing paper here. 
0  REQUEST FOR DESCRIPTION # 620
0  There is an old gun here. 
0  There is a fearsome werewolf in the room with you!
0  WOLF : TXLIM:0.300

W
0  RETURN FROM MOVE: ROOM=       5 LROOM =       3
0  TOP OF LOOP (STATEMENT 20)
0  NUMBER OF MOVES IS  21
0  ABOUT TO CALL "DES": II IS 205
0  REQUEST FOR DESCRIPTION # 205
0  You are in the dining room.
0  There is a fearsome werewolf in the room with you!
0  The werewolf attacks you with its sharp claw!
0  It severs your jugular vein and mortally wounds you!
0  WOLF : TXLIM:0.300
0  REQUEST FOR DESCRIPTION # 850
0  My, My.  You seem to have bitten the dust.
   I can attempt to reincarnate you, but I'm
   not very good at it. Should I try?
```

**Comment by Quuxplusone** [comment]
Ah, I'm starting to see what you mean. On the one hand, this could be an "original bug" — a bit of code that simply didn't do what the author intended — similar to the various misspelled messages and the way the game unconditionally prints "It is now pitch dark" when your lamp runs out of fluid (even when you are above ground). The werewolf still has _some_ chance of killing you each turn; it's just that the chance doesn't increase, because the line that increments `XLIM` is a dead write.

_However!_ It appears that many old Fortrans did save local variables in static storage by default:

- http://rsusu1.rnd.runnet.ru/develop/fortran/prof77/node156.html from 1998: "Many current Fortran systems have a simple static storage allocation scheme in which all variables are saved whether SAVE is used or not. [...] You should always take care to use the SAVE statement anywhere that its use is indicated to make your programs safe and portable."
- https://docs.oracle.com/cd/E19957-01/805-4939/6j4m0vnb7/index.html from 2010, but talking about Fortran 77: "Local variables and arrays are already static by default, predisposing the need for SAVE. However, using SAVE can ensure portability [...]"

So if VS Fortran was one of these, then I might need to add `SAVE XLIM` to the modernized code in the `patches` branch. And the same "might need" applies to literally every local variable used in any of these subroutines, ugh!

IIUC, if `XLIM` is saved between calls, then the werewolf will be _guaranteed_ to kill you for sure within 8 turns of showing up (if you don't deal with him), because `0.075 + 7*0.15 > 1.0`. This sounds plausible to me, actually.

I'll continue googling to see if I can find out whether VS Fortran implicitly saved local variables, and/or if I can find any local variables that would make the game literally unwinnable if they weren't implicitly saved. I do predict that I'll end up adding `SAVE XLIM` and perhaps a bunch more `SAVE`s  to the `patches` branch by the end of this week!

**Comment by Quuxplusone** [comment]
I see that `GGNOME` has the same structure; its dead write to `XLIM` is 0.20 instead of 0.15, meaning that if we make this change, an un-dealt-with gnome will kill the player for sure within 6 turns of showing up. (And if we _don't_ make the change, then there's no difference in deadliness between the werewolf and the gnome, which continues to seem like evidence in favor of the change.)

...Oh, and without saving `XLIM`, I guess its value on second entry to the subroutine (when `WOLF` or `GNOME` is already true) must be uninitialized, which is bad too!

So, without external evidence from googling, but _with_ internal evidence from my source-diving, I believe that we need to add `SAVE XLIM` in both `WWOLF` and `GGNOME`, and that these are the _only_ two variables in the entire program that need that patch. My timeline for making that patch has moved up to "probably by end of day today." ;)

**Comment by Quuxplusone** [comment]
- http://www.ifuap.buap.mx/manuales/profortran/UserGuide.pdf from 1999: "Absoft Fortran 77 can be made even more compatible with VS FORTRAN by using these compiler options: [...] Force all program storage to be treated as static and initialized to zero [...]"

...Jackpot!

- http://bitsavers.trailing-edge.com/pdf/ibm/370/OS_VS/fortran/GC26-3986-0_VS_FORTRAN_Prog_Ref_Feb81.pdf literally the VS-FORTRAN manual from 1981!: "Because VS FORTRAN saves these names without user action, the SAVE statement serves only as a documentation aid."

I'm convinced.

**Comment by eriktorbjorn** [comment]
Nice! I'm still surprised you got killed by the werewolf, because I was under the impression that currently `XLIM` was always 0 when it got to

```
           VAL = RDM(SEED)
           IF (VAL .GT. XLIM) GOTO 100
C GETS YOU.
                II = 1
                WRITE(6,1003)
                GOTO 900
```

But maybe I was completely wrong about that, or `VAL` can be exactly 0, or `XLIM` was undefined, not 0, or...

Anyway, sounds like you may have some material for another blog post before all of this is over. :-)

**Comment by welash** [comment]
It seems like you may have figured out the problem, but I was a little surprised that the code in rstart.f had
```
      INTEGER FUNCTION RDM(SEED)
      REAL R
      CALL RANDOM_NUMBER(R)
      RDM = R
      RETURN
      END
```
So that the function RDM returns an INTEGER instead of a REAL,  If I add -Wall to the makefile, it does give a warning
```
f77 -Wall -c rstart.f -o rstart.o
rstart.f:10:12:

   10 |       RDM = R
      |            1
Warning: Possible change of value in conversion from REAL(4) to INTEGER(4) at (1) [-Wconversion]
```
I can't remember how fortran handles this (It's been 40 years, and I only used if in college, and it was Fortran IV, not Fortran 77, so things could be different between those versions).  I was wondering if that might cause the random result to always be 0 or something.

**Comment by welash** [comment]
It also looks like you can get gfortran to use static variables for local variables with the compiler switch `-fno-automatic`

**Comment by Quuxplusone** [comment]
Re `INTEGER FUNCTION RDM`, it looks like (the inverse of) https://stackoverflow.com/a/58548821/1424877 was part of the answer: since the identifier `RDM` isn't explicitly declared anywhere else, and doesn't begin with `I` through `N`, everywhere else in the program _will_ treat it as a `REAL FUNCTION` regardless of how it's declared in `rstart.f`. As to what exactly was happening on the line `RDM = R` (where `R` is implicitly real and `RDM` was declared `INTEGER FUNCTION` and gfortran complained about a type conversion), I don't know enough Fortran either. Empirically, I printed out some results of `RDM(SEED)` in a loop in main, and they looked to be in the correct range no matter whether `rstart.f` had `INTEGER FUNCTION RDM` or `REAL FUNCTION RDM`.
Anyway, since this was code I myself added in the `patches` branch, I have now corrected it by changing to `REAL FUNCTION RDM`.
